ribadstore.cc

```diff
@@ -176,13 +221,50 @@ namespace ns3
                 std::string ad(ss.str());
 
                 // * printout the received packet body
-                NS_LOG_INFO(ad);
-                
+                RIB *rib = (RIB *)(this->parent_ctx);
+                // NS_LOG_INFO("I am ribadstore at " << Ipv4Address::ConvertFrom(rib->my_addr));
+                NS_LOG_INFO("" << Ipv4Address::ConvertFrom(rib->my_addr) <<  " received: " << ad);
+
                 if (ad == "GIVEPEERS"){
                     Simulator::ScheduleNow(&RIBAdStore::SendPeers, this, socket, from);
                 }else{
-                    db.insert(ad);
+                    // * deserialize the advertisement packet
+                    NameDBEntry* advertised_entry = NameDBEntry::FromAdvertisementStr(ad);
+                    if (advertised_entry == nullptr) {
+                        NS_LOG_ERROR("Cannot parse ads: " << ad);
+                        // std::abort();
+                        continue;
+                    }
+                    bool updated = UpdateNameCache(advertised_entry);
                     NS_LOG_INFO("Number of ads: " << db.size());
+
+                    // if new udpate, broadcast to peers if r_transitivity allows
+                    if (updated) {
+                        RIB *rib = (RIB *)(this->parent_ctx);
+                        // check the r_transitivity before advertise
+                        if (advertised_entry->r_transitivity > 1) {
+                            -- advertised_entry->r_transitivity;
+                            std::string serialized = advertised_entry->ToAdvertisementStr();
+                            // need to add back 1 because the decrease was directly modifying the entry itself
+                            ++ advertised_entry->r_transitivity;
+
+                            for (auto addr : rib->peers) {
+                                //   cases not to forware:
+                                //     1. the destination is what this ads came from
+                                //     2. the destination is the origin AS
+                                //     3. .. 
+                                Address dest_socket = InetSocketAddress(Ipv4Address::ConvertFrom(addr), RIBADSTORE_PORT);
```

Comment:
Learning: ns3 socket actually need to create socket address before sent out packets. 
Address is the parent class of Ipv4Address. 
Address ---->> Ipv4Address: Just call Ipv4Address::ConvertFrom()
Ipv4Address ---->>Address: Just simple cast